stdlib/src/builtin/string.mojo

```diff
@@ -585,11 +585,15 @@ fn isspace(c: UInt8) -> Bool:
         True if the character is one of the whitespace characters listed above, otherwise False.
     """
 
-    alias ord_space = ord(" ")
-    alias ord_tab = ord("\t")
-    alias ord_carriage_return = ord("\r")
-
-    return c == ord_space or ord_tab <= int(c) <= ord_carriage_return
+    alias c_spaces = (
+        ord(" "),
+        ord("\t"),
+        ord("\n"),
+        ord("\r"),
+        ord("\f"),
+        ord("\v"),
+    )
+    return int(c) in c_spaces
```

Comment:
If possible, we should just do this with a 256 element jump table. A hidden top-level alias could do this.

Reply:
> do this with a 256 element jump table. A hidden top-level alias could do this

Do you have any tips on how to approach that? all examples I've found are with switch in C, or talk about logic truth tables (or FPGAs which have nothing to do here)...

Won't it be the same performance by doing `Tuple[UInt8](...).__contains__` ?

And another question, standalone this works fine but in stringref when using the `UnsafePointer[SIMD[uint8, 1], 0]` it doesn't (tests fail). Is there some reason for that? will this approach even make a performance difference?
```mojo
fn main():
    alias spaces = SIMD[DType.uint8, 8](
        ord(" "),
        ord("\t"),
        ord("\n"),
        ord("\r"),
        ord("\f"),
        ord("\v"),
        ord("\v"),
        ord("\v"),
    )
    var item1: UInt8 = ord(" ")
    print((spaces & item1).cast[DType.bool]().reduce_or())
    var item2 = UnsafePointer[SIMD[DType.uint8, 1], 0](UInt8(ord(" ")))
    print((item2[0] & spaces).cast[DType.bool]().reduce_or())
```

Reply:
Okay, I think what we need is a module level variable, not an alias. I think something like this might work:
```mojo
alias _SpacesTableType = InlineArray[UInt8, 256]

fn _get_spaces_table() -> _SpacesTableType:
    var table = _SpacesTableType(0)
    ...  # manually set locations associated with space types to 1, e.g. `table[ord(" ")] = 1` (no need to loop over these, just write them out)
    return table

var _SPACES_TABLE = _get_spaces_table()

fn isspace(char: UInt8) -> Bool:
   return _SPACES_TABLE[char]

```
This is an aggressive tradeoff: takes 256 bytes of memory but the lookup is very fast. I don't know if we can align this table in memory, but maybe we should.

There is a chance this won't work at all (because globals are not really baked out in mojo), but it's worth pursuing.

Reply:
I'm not sure if I like using so much memory without explicit command from the developer, is there some way we can make this opt in? For micro-controller development 256 bytes counts for a lot. And also other methods like `splitlines()` will need to define their own character combinations anyway, since you have stuff like windows `\r\n`.

Reply:
We are very far from being able to run in any memory constrained environment and Python itself employs this jump table optimization for performance. When we start looking at MCUs and other embedded system, we will be able to specialize the function for different targets. For the systems we run on today, however, this optimization is needed for performance; looping through collections of characters/strings and looking for spaces is common, and it has to be fast.

Reply:
> probably you can just compare the underlying List

ok done, used memcmp

What do you think of this approach? it works, I was doing the logic wrong earlier.
```mojo
fn isspace(c: UInt8) -> Bool:
    alias spaces = SIMD[DType.uint8, 8](
        ord(" "),
        ord("\t"),
        ord("\n"),
        ord("\r"),
        ord("\f"),
        ord("\v"),
        ord("\v"),
        ord("\v"),
    )
    return (spaces ^ c).reduce_min() == 0
```

> We might have to employ similar optimizations there. Let's cross that bridge when we get there.

we might just use the same LUT and do some hashing to get bigger sequences in there (?). It would be kinda cheating but for example:

\x85 or 120 56 53 or 0x78 0x38 0x35 is NEXT LINE (NEL) character
\x1c or 92 120 49 99 or 0x5c 0x78 0x31 0x63 is INFORMATION SEPARATOR FOUR character
\x1e or 92 120 49 101 or 0x5c 0x78 0x31 0x65 is INFORMATION SEPARATOR TWO character

we could just grab the last byte of each and `|` it with a `0b10000000` (though other multi byte sequences might end up getting returned as true)

Also I've been thinking of how will we deal in other algos with the fact that some bytes might have `11xxxx...` as their first 2 bits and we should look for the nex byte while `10xxx...` is their first 2 bits. Or if we are already reading the first byte then we could just directly do
```mojo
var num_bytes = 1
if b >> 7 != 0:
    num_bytes = (bit_reverse(b) & 0b1111)
```
and we get the number of bytes to grab next. I'm worried for the performance either way.

Reply:
Can you please just implement the LUT for length 1? It is a known fast implementation with little implementation complexity. Right now, we do not care about 256 bytes, and simd reduction may or may not be fast depending on the machine, the reduce implementation, and the compiler optimizations we can do. Once we have that, we can start a separate discussion on how to optimize these with different tradeoffs on different platforms, but only if needed.